File: code/python/06-frameworks/autogen_design_team.py
"""AutoGen product design team: conversational multi-agent design.

Implements a five-agent group chat using AutoGen (Microsoft):

    ProductManager ─┐
    Designer        │
    Engineer        ├─→ GroupChatManager → consensus → DesignSpec
    Critic          │
    UserProxy       ┘

Solutions emerge from conversation rather than from a predefined task list.
That is the fundamental difference from CrewAI: here, no agent "plans" the
discussion — the dialogue itself produces the outcome.

When AutoGen is not installed a custom conversational loop provides the same
pattern using only the OpenAI SDK, so you can compare the two approaches.

Run:
    python autogen_design_team.py

See: docs/06-frameworks-in-practice/03-crewai-autogen.md
"""

from __future__ import annotations

import os
import re
import time
from dataclasses import dataclass, field
from typing import Optional

from openai import OpenAI

# ---------------------------------------------------------------------------
# Optional AutoGen import — pyautogen (v0.2/v0.3 API)
# ---------------------------------------------------------------------------
# AutoGen has gone through significant API changes.  We target the stable
# v0.2/v0.3 interface published as ``pyautogen``.  A v0.4+ port would use
# ``autogen_agentchat`` but the group-chat concepts are identical.
# ---------------------------------------------------------------------------

try:
    import autogen  # type: ignore[import]
    from autogen import AssistantAgent, UserProxyAgent, GroupChat, GroupChatManager  # type: ignore[import]

    AUTOGEN_AVAILABLE = True
except ImportError:
    AUTOGEN_AVAILABLE = False

# ---------------------------------------------------------------------------
# Shared LLM client (also used by the from-scratch fallback)
# ---------------------------------------------------------------------------

_client: Optional[OpenAI] = None


def _get_client() -> OpenAI:
    global _client
    if _client is None:
        _client = OpenAI(api_key=os.environ.get("OPENAI_API_KEY", ""))
    return _client


# ---------------------------------------------------------------------------
# Result dataclasses
# ---------------------------------------------------------------------------


@dataclass
class ConversationAnalysis:
    """Statistical breakdown of a completed group-chat conversation."""

    speaker_turns: dict[str, int] = field(default_factory=dict)
    productive_rounds: int = 0
    circular_rounds: int = 0
    decisions: list[str] = field(default_factory=list)
    deadlocks_detected: int = 0
    total_rounds: int = 0


@dataclass
class DesignResult:
    """Structured output from any product design team implementation."""

    requirement: str
    final_spec: str = ""
    conversation_history: list[dict] = field(default_factory=list)
    rounds_taken: int = 0
    decisions_made: list[str] = field(default_factory=list)
    execution_time: float = 0.0
    token_usage: int = 0


# ---------------------------------------------------------------------------
# ProductDesignTeam — AutoGen implementation
# ---------------------------------------------------------------------------


class ProductDesignTeam:
    """An AutoGen group chat that designs a product feature through conversation.

    Agents:
    - ProductManager: Defines requirements and priorities
    - Designer: Proposes UX solutions
    - Engineer: Evaluates technical feasibility
    - Critic: Challenges assumptions and finds edge cases
    - UserProxy: Represents the user; can ask for human input

    The conversation proceeds until the team reaches consensus on a spec,
    or until max_rounds is exhausted.
    """

    TERMINATE_SIGNAL = "DESIGN_APPROVED"

    def __init__(self, model: str = "gpt-4o") -> None:
        if not AUTOGEN_AVAILABLE:
            raise ImportError(
                "pyautogen is required for ProductDesignTeam. "
                "Install with: pip install pyautogen"
            )
        self.llm_config: dict = {
            "config_list": [
                {
                    "model": model,
                    "api_key": os.environ.get("OPENAI_API_KEY", ""),
                }
            ],
            "temperature": 0.7,
        }
        self.agents = self._create_agents()

    def _create_agents(self) -> dict:
        """Create all agents with distinct system messages.

        Each agent has a clear personality and a well-defined area of expertise.
        This is what shapes the conversation: roles create productive tension.
        """
        product_manager = AssistantAgent(
            name="ProductManager",
            llm_config=self.llm_config,
            system_message=(
                "You are an experienced product manager. Your job is to clarify "
                "requirements, set priorities, and keep the team focused on user "
                "value. When the team proposes something technically interesting "
                "but user-unfriendly, push back. When the team reaches a decision, "
                "summarise it clearly. If the design is complete and the team agrees, "
                f"end your message with '{self.TERMINATE_SIGNAL}'."
            ),
        )

        designer = AssistantAgent(
            name="Designer",
            llm_config=self.llm_config,
            system_message=(
                "You are a senior UX designer. You propose user interface solutions "
                "grounded in usability principles and accessibility standards. You "
                "think about edge cases: what happens when two users edit the same "
                "thing simultaneously? What does the error state look like? Always "
                "sketch the happy path first, then cover failure modes."
            ),
        )

        engineer = AssistantAgent(
            name="Engineer",
            llm_config=self.llm_config,
            system_message=(
                "You are a senior software engineer. You evaluate design proposals "
                "for technical feasibility, performance implications, and security "
                "risks. Be specific: if something requires O(n²) syncing or "
                "introduces a race condition, say so. Propose concrete technical "
                "solutions, not just objections."
            ),
        )

        critic = AssistantAgent(
            name="Critic",
            llm_config=self.llm_config,
            system_message=(
                "You are a rigorous product critic. Your job is to find flaws that "
                "the team has missed: edge cases, unmet user needs, inconsistencies "
                "between requirements and proposed solutions. Be constructive — "
                "your goal is a better product, not winning the argument. Prioritise "
                "your objections; not everything is equally important."
            ),
        )

        user_proxy = UserProxyAgent(
            name="UserProxy",
            human_input_mode="NEVER",  # Fully automated; change to "TERMINATE" for interactive
            max_consecutive_auto_reply=1,
            is_termination_msg=lambda msg: self.TERMINATE_SIGNAL in msg.get("content", ""),
            code_execution_config=False,
        )

        return {
            "product_manager": product_manager,
            "designer": designer,
            "engineer": engineer,
            "critic": critic,
            "user_proxy": user_proxy,
        }

    def design_feature(
        self, requirement: str, max_rounds: int = 20
    ) -> DesignResult:
        """Run the design conversation and return structured results.

        Returns:
            DesignResult with requirement, final_spec, conversation_history,
            rounds_taken, decisions_made, execution_time, token_usage.
        """
        agents = self.agents
        groupchat = GroupChat(
            agents=[
                agents["user_proxy"],
                agents["product_manager"],
                agents["designer"],
                agents["engineer"],
                agents["critic"],
            ],
            messages=[],
            max_round=max_rounds,
            speaker_selection_method="round_robin",
        )
        manager = GroupChatManager(
            groupchat=groupchat,
            llm_config=self.llm_config,
        )

        start = time.perf_counter()
        agents["user_proxy"].initiate_chat(
            manager,
            message=(
                f"We need to design the following product feature.\n\n"
                f"REQUIREMENT: {requirement}\n\n"
                "Please work through the design together. ProductManager: start "
                "by clarifying the requirements. Designer: propose solutions. "
                "Engineer: evaluate feasibility. Critic: challenge assumptions. "
                "Iterate until you reach consensus on a complete design spec."
            ),
        )
        elapsed = time.perf_counter() - start

        history = groupchat.messages
        decisions = _extract_decisions(history)
        final_spec = _build_final_spec(history, decisions)

        return DesignResult(
            requirement=requirement,
            final_spec=final_spec,
            conversation_history=[
                {"role": m.get("name", "unknown"), "content": m.get("content", "")}
                for m in history
            ],
            rounds_taken=len(history),
            decisions_made=decisions,
            execution_time=elapsed,
        )

    def analyze_conversation(self, history: list[dict]) -> ConversationAnalysis:
        """Analyze a conversation for participation, productivity, and decisions.

        Returns:
            ConversationAnalysis with speaker turns, productive vs. circular
            rounds, decisions made, deadlocks, and total rounds.
        """
        speaker_turns: dict[str, int] = {}
        decisions: list[str] = []
        seen_topics: set[str] = set()
        circular_count = 0
        productive_count = 0

        for msg in history:
            speaker = msg.get("role", msg.get("name", "unknown"))
            content = msg.get("content", "")

            speaker_turns[speaker] = speaker_turns.get(speaker, 0) + 1

            # Heuristic: detect decisions (ProductManager summaries, approvals)
            if "DESIGN_APPROVED" in content or re.search(
                r"\bdecid(ed|e|ing)\b|\bagreed?\b|\bwe will\b", content, re.I
            ):
                # Extract the first sentence as the decision
                first_sentence = content.split(".")[0].strip()
                if first_sentence and first_sentence not in decisions:
                    decisions.append(first_sentence[:120])

            # Heuristic: circular if the same key phrase appears again
            key_phrase = re.sub(r"\W+", " ", content[:60].lower()).strip()
            if key_phrase in seen_topics:
                circular_count += 1
            else:
                seen_topics.add(key_phrase)
                productive_count += 1

        deadlocks = max(0, circular_count - 2)  # 1-2 repeats are normal debate

        return ConversationAnalysis(
            speaker_turns=speaker_turns,
            productive_rounds=productive_count,
            circular_rounds=circular_count,
            decisions=decisions,
            deadlocks_detected=deadlocks,
            total_rounds=len(history),
        )


# ---------------------------------------------------------------------------
# CustomConversationalTeam — from-scratch fallback (no AutoGen required)
# ---------------------------------------------------------------------------


class CustomConversationalTeam:
    """Same five-agent conversation pattern using only the OpenAI SDK.

    Demonstrates that AutoGen's conversational model is conceptually simple:
    each agent is a system prompt, messages accumulate in shared history,
    and the manager decides who speaks next.
    """

    TERMINATE_SIGNAL = "DESIGN_APPROVED"

    def __init__(self, model: str = "gpt-4o") -> None:
        self.model = model
        self.client = _get_client()
        self._system_prompts = self._build_system_prompts()

    def _build_system_prompts(self) -> dict[str, str]:
        return {
            "ProductManager": (
                "You are an experienced product manager. Clarify requirements, "
                "set priorities, and keep the team focused on user value. When "
                f"the design is complete and the team agrees, say '{self.TERMINATE_SIGNAL}'."
            ),
            "Designer": (
                "You are a senior UX designer. Propose user interface solutions "
                "grounded in usability principles. Cover the happy path first, "
                "then failure modes and edge cases."
            ),
            "Engineer": (
                "You are a senior software engineer. Evaluate design proposals for "
                "technical feasibility, performance, and security. Be specific about "
                "risks and propose concrete solutions."
            ),
            "Critic": (
                "You are a rigorous product critic. Find edge cases, inconsistencies, "
                "and unmet user needs the team has missed. Be constructive and "
                "prioritise your objections."
            ),
        }

    def design_feature(
        self, requirement: str, max_rounds: int = 20
    ) -> DesignResult:
        """Run the design conversation using a simple round-robin loop."""
        turn_order = ["ProductManager", "Designer", "Engineer", "Critic"]
        conversation: list[dict] = []
        token_total = 0
        decisions: list[str] = []

        # Seed the conversation
        conversation.append({
            "role": "user",
            "content": (
                f"We need to design the following product feature.\n\n"
                f"REQUIREMENT: {requirement}\n\n"
                "ProductManager: start by clarifying the requirements."
            ),
        })

        start = time.perf_counter()
        for round_num in range(max_rounds):
            speaker = turn_order[round_num % len(turn_order)]
            sys_prompt = self._system_prompts[speaker]

            messages_for_call = [{"role": "system", "content": sys_prompt}]
            # Include the last 8 turns for context (sliding window)
            messages_for_call.extend(conversation[-8:])

            response = self.client.chat.completions.create(
                model=self.model,
                messages=messages_for_call,
            )
            reply = response.choices[0].message.content or ""
            token_total += response.usage.total_tokens if response.usage else 0

            conversation.append({"role": "assistant", "name": speaker, "content": reply})

            # Capture decisions
            if re.search(r"\bdecid(ed|e|ing)\b|\bagreed?\b|\bwe will\b", reply, re.I):
                sentence = reply.split(".")[0].strip()
                if sentence not in decisions:
                    decisions.append(sentence[:120])

            # Termination check
            if self.TERMINATE_SIGNAL in reply:
                break

        elapsed = time.perf_counter() - start
        final_spec = _build_final_spec(conversation, decisions)

        return DesignResult(
            requirement=requirement,
            final_spec=final_spec,
            conversation_history=[
                {
                    "role": m.get("name", m.get("role", "unknown")),
                    "content": m.get("content", ""),
                }
                for m in conversation
            ],
            rounds_taken=len(conversation),
            decisions_made=decisions,
            execution_time=elapsed,
            token_usage=token_total,
        )

    def analyze_conversation(self, history: list[dict]) -> ConversationAnalysis:
        """Delegate to the same analysis logic as ProductDesignTeam."""
        _team = ProductDesignTeam.__new__(ProductDesignTeam)
        return _team.analyze_conversation(history)


# ---------------------------------------------------------------------------
# Shared helpers
# ---------------------------------------------------------------------------


def _extract_decisions(history: list[dict]) -> list[str]:
    """Pull decision-like sentences from conversation history."""
    decisions: list[str] = []
    for msg in history:
        content = msg.get("content", "")
        if re.search(r"\bdecid(ed|e|ing)\b|\bagreed?\b|\bwe will\b|\bfinal spec\b", content, re.I):
            sentence = content.split(".")[0].strip()
            if sentence and sentence not in decisions:
                decisions.append(sentence[:120])
    return decisions


def _build_final_spec(history: list[dict], decisions: list[str]) -> str:
    """Construct a final spec string from the last few messages and decisions."""
    if not history:
        return "(no conversation recorded)"

    # Look for an APPROVED message first
    for msg in reversed(history):
        if "DESIGN_APPROVED" in msg.get("content", ""):
            return msg["content"]

    # Fallback: summarise the last message + decisions
    last_content = history[-1].get("content", "")
    decision_text = "\n".join(f"• {d}" for d in decisions) if decisions else "(none)"
    return (
        f"DESIGN SPEC (reconstructed from conversation)\n\n"
        f"Key decisions:\n{decision_text}\n\n"
        f"Final team message:\n{last_content}"
    )


# ---------------------------------------------------------------------------
# Demo
# ---------------------------------------------------------------------------


def _divider(title: str, width: int = 60) -> None:
    print(f"\n{'─'*width}")
    print(f"  {title}")
    print(f"{'─'*width}")


def _preview(text: str, max_chars: int = 400) -> str:
    if len(text) <= max_chars:
        return text
    return text[:max_chars] + f"\n  … [{len(text) - max_chars} more chars]"


if __name__ == "__main__":
    REQUIREMENT = "Add collaborative editing to our document editor"

    if AUTOGEN_AVAILABLE:
        print("AutoGen detected. Running ProductDesignTeam.\n")
        team: ProductDesignTeam | CustomConversationalTeam = ProductDesignTeam()
    else:
        print("AutoGen not installed. Running CustomConversationalTeam fallback.")
        print("To enable AutoGen: pip install pyautogen\n")
        team = CustomConversationalTeam()

    result = team.design_feature(REQUIREMENT, max_rounds=12)

    _divider("REQUIREMENT")
    print(f"  {result.requirement}")

    _divider("CONVERSATION")
    for i, turn in enumerate(result.conversation_history):
        speaker = turn.get("role", "unknown")
        content = turn.get("content", "")
        print(f"\n[{i+1}] {speaker.upper()}")
        print(_preview(content, max_chars=300))

    _divider("FINAL SPEC")
    print(result.final_spec)

    _divider("CONVERSATION ANALYSIS")
    analysis = team.analyze_conversation(result.conversation_history)
    print(f"  Total rounds    : {analysis.total_rounds}")
    print(f"  Productive      : {analysis.productive_rounds}")
    print(f"  Circular        : {analysis.circular_rounds}")
    print(f"  Deadlocks       : {analysis.deadlocks_detected}")
    print(f"\n  Speaker turns:")
    for speaker, turns in sorted(analysis.speaker_turns.items(), key=lambda x: -x[1]):
        print(f"    {speaker:<20} {turns} turns")
    print(f"\n  Decisions made  : {len(analysis.decisions)}")
    for d in analysis.decisions[:5]:
        print(f"    • {d}")

    _divider("WHERE THE CRITIC IMPROVED THE DESIGN")
    critic_turns = [
        t for t in result.conversation_history
        if t.get("role", "").lower() == "critic"
    ]
    if critic_turns:
        print(f"\n  Critic spoke {len(critic_turns)} time(s). Key contribution:")
        print(_preview(critic_turns[0].get("content", ""), max_chars=400))
    else:
        print("  (Critic did not contribute in this run)")

    print(f"\n  Execution time  : {result.execution_time:.1f}s")
    print(f"  Tokens used     : {result.token_usage:,}")

File: code/go/06-frameworks/autogen_design_team.go
// autogen_design_team.go — AutoGen-equivalent conversational design team in Go.
//
// Python AutoGen has no official Go port. This file implements the same
// CONCEPT — a group-chat orchestrator that routes messages between
// specialised agents until consensus is reached — using Go-native code.
//
// Agents in the team:
//   - ProductManager  : defines user requirements and acceptance criteria
//   - Architect       : designs high-level system structure
//   - SecurityReviewer: identifies risks and enforces least-privilege principle
//
// The orchestrator runs up to maxConversationRuns rounds; agents signal
// "APPROVED" when the spec is complete. If no consensus is reached, the
// best draft after the final round is returned.
//
// Run:
//
//	go run autogen_design_team.go hybrid_rag_agent.go langgraph_alternative.go \
//	           multi_agent_from_scratch.go framework_comparison.go framework_advisor.go \
//	           framework_extraction.go over_engineering_detector.go langsmith_tracer.go
//
// See: docs/06-frameworks-in-practice/03-crewai-autogen.md
package main

import (
	"context"
	"fmt"
	"strings"
	"time"

	openai "github.com/sashabaranov/go-openai"
)

// ---------------------------------------------------------------------------
// Result types
// ---------------------------------------------------------------------------

// DesignResult holds the output of a conversational design session.
type DesignResult struct {
	Requirement         string
	FinalSpec           string
	ConversationHistory []map[string]string // keys: role, name, content
	RoundsTaken         int
	DecisionsMade       []string
	ExecutionTimeSec    float64
	TokenUsage          int
}

// DesignConversationAnalysis analyses the quality of a design conversation.
type DesignConversationAnalysis struct {
	SpeakerTurns      map[string]int
	ProductiveRounds  int
	CircularRounds    int
	Decisions         []string
	DeadlocksDetected int
	TotalRounds       int
}

// ---------------------------------------------------------------------------
// Design team orchestrator
// ---------------------------------------------------------------------------

// dtAgentSpec defines one participant in the design conversation.
type dtAgentSpec struct {
	Name      string
	Role      string
	SystemMsg string
}

// dtDesignAgents returns the three agents of the team.
func dtDesignAgents() []dtAgentSpec {
	return []dtAgentSpec{
		{
			Name: "ProductManager",
			Role: "product manager",
			SystemMsg: "You are a product manager. Your job is to:\n" +
				"1. Define clear user requirements and acceptance criteria.\n" +
				"2. Prioritise features (MUST/SHOULD/NICE-TO-HAVE).\n" +
				"3. Challenge over-engineering.\n" +
				"Respond concisely. When the spec is complete write 'APPROVED' on the last line.",
		},
		{
			Name: "Architect",
			Role: "solutions architect",
			SystemMsg: "You are a solutions architect. Your job is to:\n" +
				"1. Design a scalable, maintainable system structure.\n" +
				"2. Choose appropriate data stores and APIs.\n" +
				"3. Identify integration points and dependencies.\n" +
				"Respond concisely. When the design is finalized write 'APPROVED' on the last line.",
		},
		{
			Name: "SecurityReviewer",
			Role: "application security engineer",
			SystemMsg: "You are an application security engineer. Your job is to:\n" +
				"1. Identify security risks (OWASP Top-10, data exposure, authentication gaps).\n" +
				"2. Enforce least-privilege access and data minimisation.\n" +
				"3. Require mitigations before approving.\n" +
				"Respond concisely. When all risks are mitigated write 'APPROVED' on the last line.",
		},
	}
}

// dtRunAgent calls the LLM for a single agent turn.
func dtRunAgent(ctx context.Context, client *openai.Client, spec dtAgentSpec, history []openai.ChatCompletionMessage) (string, int, error) {
	msgs := append([]openai.ChatCompletionMessage{
		{Role: openai.ChatMessageRoleSystem, Content: spec.SystemMsg},
	}, history...)

	resp, err := client.CreateChatCompletion(ctx, openai.ChatCompletionRequest{
		Model:     "gpt-4o-mini",
		Messages:  msgs,
		MaxTokens: 400,
	})
	if err != nil {
		return "", 0, err
	}
	return resp.Choices[0].Message.Content, resp.Usage.TotalTokens, nil
}

// dtRunDesignSession orchestrates the group-chat conversation.
func dtRunDesignSession(ctx context.Context, client *openai.Client, requirement string) (DesignResult, error) {
	t0 := time.Now()
	maxRounds := 4
	agents := dtDesignAgents()

	result := DesignResult{Requirement: requirement}

	// history for the LLM context (grows with each turn)
	history := []openai.ChatCompletionMessage{
		{Role: openai.ChatMessageRoleUser, Content: "Design requirement:\n" + requirement},
	}
	// readable record for the result
	var convHistory []map[string]string
	convHistory = append(convHistory, map[string]string{
		"role": "user", "name": "Initiator", "content": requirement,
	})

	approvals := make(map[string]bool)

	for round := 1; round <= maxRounds; round++ {
		result.RoundsTaken = round
		allApproved := true

		for _, agent := range agents {
			if approvals[agent.Name] {
				continue
			}

			reply, tokens, err := dtRunAgent(ctx, client, agent, history)
			if err != nil {
				return result, fmt.Errorf("agent %s round %d: %w", agent.Name, round, err)
			}
			result.TokenUsage += tokens

			convHistory = append(convHistory, map[string]string{
				"role": "assistant", "name": agent.Name, "content": reply,
			})
			history = append(history, openai.ChatCompletionMessage{
				Role:    openai.ChatMessageRoleAssistant,
				Content: fmt.Sprintf("[%s]: %s", agent.Name, reply),
			})

			if strings.Contains(strings.ToUpper(reply), "APPROVED") {
				approvals[agent.Name] = true
			} else {
				allApproved = false
			}
		}

		if allApproved {
			break
		}
	}

	result.ConversationHistory = convHistory
	result.DecisionsMade = dtExtractDecisions(convHistory)
	result.FinalSpec = dtBuildFinalSpec(convHistory, result.DecisionsMade)
	result.ExecutionTimeSec = time.Since(t0).Seconds()

	return result, nil
}

// ---------------------------------------------------------------------------
// Analysis helpers
// ---------------------------------------------------------------------------

// dtExtractDecisions scans the conversation for lines that look like decisions.
func dtExtractDecisions(history []map[string]string) []string {
	decisionIndicators := []string{
		"we should", "we will", "we need", "must use", "must not",
		"decision:", "agreed:", "require", "will use", "will not",
	}
	seen := make(map[string]bool)
	var decisions []string

	for _, msg := range history {
		if msg["role"] != "assistant" {
			continue
		}
		for _, line := range strings.Split(msg["content"], "\n") {
			lineLower := strings.ToLower(strings.TrimSpace(line))
			if lineLower == "" || lineLower == "approved" {
				continue
			}
			for _, indicator := range decisionIndicators {
				if strings.Contains(lineLower, indicator) {
					key := strings.ToLower(line[:min(len(line), 80)])
					if !seen[key] {
						seen[key] = true
						trimmed := strings.TrimSpace(line)
						if len(trimmed) > 120 {
							trimmed = trimmed[:120] + "…"
						}
						decisions = append(decisions, trimmed)
					}
					break
				}
			}
		}
	}
	return decisions
}

// dtBuildFinalSpec assembles the specification from the last assistant message
// in the conversation that contains more than one sentence.
func dtBuildFinalSpec(history []map[string]string, decisions []string) string {
	var lastSubstantive string
	for _, msg := range history {
		if msg["role"] == "assistant" {
			content := strings.TrimSpace(msg["content"])
			// Remove "APPROVED" trailing line
			content = strings.TrimSuffix(strings.TrimSpace(strings.ToUpper(content)), "APPROVED")
			content = strings.TrimSpace(msg["content"])
			if len(content) > 100 {
				lastSubstantive = content
			}
		}
	}

	var sb strings.Builder
	sb.WriteString("=== DESIGN SPECIFICATION ===\n\n")
	if lastSubstantive != "" {
		sb.WriteString(lastSubstantive)
		sb.WriteString("\n\n")
	}
	if len(decisions) > 0 {
		sb.WriteString("=== KEY DECISIONS ===\n")
		for i, d := range decisions {
			sb.WriteString(fmt.Sprintf("%d. %s\n", i+1, d))
		}
	}
	return sb.String()
}

// analyzeDesignConversation returns statistics about the conversation quality.
func analyzeDesignConversation(history []map[string]string) DesignConversationAnalysis {
	analysis := DesignConversationAnalysis{
		SpeakerTurns: make(map[string]int),
		Decisions:    dtExtractDecisions(history),
	}
	analysis.TotalRounds = len(history)

	prevContent := make(map[string]string)
	for _, msg := range history {
		if msg["role"] != "assistant" {
			continue
		}
		name := msg["name"]
		analysis.SpeakerTurns[name]++

		content := strings.TrimSpace(msg["content"])
		if prev, ok := prevContent[name]; ok {
			// Circular if the message is very similar to the previous one from the same agent
			if len(content) > 0 && len(prev) > 0 {
				// Simple overlap check
				shorter, longer := prev, content
				if len(shorter) > len(longer) {
					shorter, longer = longer, shorter
				}
				if len(shorter) > 20 && strings.Contains(strings.ToLower(longer), strings.ToLower(shorter[:20])) {
					analysis.CircularRounds++
				}
			}
		} else {
			analysis.ProductiveRounds++
		}
		prevContent[name] = content
	}
	return analysis
}

// ---------------------------------------------------------------------------
// Demo
// ---------------------------------------------------------------------------

// runAutogenDesignTeamDemo is the demo entry point (no main()).
func runAutogenDesignTeamDemo(ctx context.Context, client *openai.Client) {
	requirements := []string{
		"Build a real-time dashboard that shows live KPIs from our PostgreSQL database. " +
			"The dashboard needs to handle 1000 concurrent viewers with sub-second refresh.",
		"Create a customer-facing AI chatbot for product support. " +
			"It must answer questions about our product catalogue and escalate complex issues to human agents.",
	}

	for i, req := range requirements {
		fmt.Printf("\n%s\n  DESIGN SESSION %d\n%s\n\n", strings.Repeat("═", 60), i+1, strings.Repeat("═", 60))
		fmt.Printf("Requirement:\n  %s\n\n", req)

		result, err := dtRunDesignSession(ctx, client, req)
		if err != nil {
			fmt.Printf("Error: %v\n", err)
			continue
		}

		fmt.Printf("Rounds taken    : %d\n", result.RoundsTaken)
		fmt.Printf("Tokens used     : %d\n", result.TokenUsage)
		fmt.Printf("Execution time  : %.2fs\n", result.ExecutionTimeSec)
		fmt.Printf("Decisions made  : %d\n", len(result.DecisionsMade))

		if len(result.DecisionsMade) > 0 {
			fmt.Printf("\nKey Decisions:\n")
			for j, d := range result.DecisionsMade {
				fmt.Printf("  %d. %s\n", j+1, d)
			}
		}

		fmt.Printf("\nFinal Spec (preview):\n")
		spec := result.FinalSpec
		if len(spec) > 500 {
			spec = spec[:500] + "\n…(truncated)"
		}
		fmt.Printf("%s\n", spec)

		analysis := analyzeDesignConversation(result.ConversationHistory)
		fmt.Printf("\nConversation Analysis:\n")
		fmt.Printf("  Speaker turns   : %v\n", analysis.SpeakerTurns)
		fmt.Printf("  Productive turns: %d\n", analysis.ProductiveRounds)
		fmt.Printf("  Circular turns  : %d\n", analysis.CircularRounds)
		fmt.Printf("  Decisions found : %d\n", len(analysis.Decisions))
	}
}
